File: hw1/hw1.java
/**
 * The hw1 class is a direct port of hw1.c to java.
 * As you already know in java when you pass literal strings like
 * <P>
 *   writeline("a literal string\n", stream);
 * <P>
 * in C is considered a char[], but in java it's automatically
 * converted and treated as a String object.  Therefore 
 * the function writeline accepts literal strings and 
 * String types.
 */

import java.io.*;      // System.out

public class hw1 {
	private static final long COUNT = 16;	// number of hex digits to display
	private static final long DECIMAL = 10;	// to indicate base 10
	private static final long HEX = 16;		// to indicate base 16
	private static final int ADD = 12;	// to add to the var in the main loop

	private static final char digits[] = 	// used for ASCII conversion
		new String("0123456789ABCDEFGHIJKLMNOPQRSTUVWXYZ").toCharArray();
	private static long hexCounter = 0; // counter for the number hex digits


	/**
	* Takes in a positive number and displays in a given base.
        *
        * @param    Numeric value to be displayed.
        * @param    Base to used to display number.
        * @param    Where to display, likely System.out or System.err.
        */
	private static void baseout (long number, long base, PrintStream stream) {
		// YOUR CODE GOES HERE
		char[] arr = new char[(int)COUNT];

		int index = 0;
		int remainder = 0;
		do
		{
			remainder = (int)number % (int)base;
			arr[index] = digits[remainder];
			number /= base;
			index++;
		} while (number != 0);

		while (index >= 0)
		{
			fputc(arr[index], stream);
			index--;
		}
	}


        /**
        * Takes in a positive number and displays it in decimal.
        *
        * @param    positive numeric value to be displayed.
        * @param    Where to display, likely System.out or System.err.
        */
	public static void decout (long number, PrintStream stream) {
		// YOUR CODE GOES HERE
		baseout (number, DECIMAL, stream);
	}


        /**
        * Displays a single character.
        *
        * @param    Character to display.
        * @param    Where to display, likely System.out or System.err.
        */
        public static void fputc(char CharToDisp, PrintStream stream) {

                // Print a single character.
                stream.print (CharToDisp);   

                // Flush the system.out buffer, now. 
                stream.flush ();
        }


        /**
        * Takes in a positive number and displays it in hex.
        *
        * @param    A positive numeric value to be displayed in hex.
        * @param    Where to display, likely System.out or System.err.
        */
        public static void hexout (long number, PrintStream stream) {

                // Output "0x" for hexidecimal.
                writeline ("0x", stream);
                baseout (number, HEX, stream);
        }


        /**
        * Prints out a newline character.
        * @param    Where to display, likely System.out or System.err.
        *
        */
        public static void newline ( PrintStream stream ) {
		// YOUR CODE GOES HERE
		fputc('\n', stream);
        }


        /**
        * Prints out a string.
        *
        * @param    A string to print out.
        * @param    Where to display, likely System.out or System.err.
        * @return     <code>int</code> The length of the string.
        */
        public static long writeline (String message, PrintStream stream) {
                // YOUR CODE GOES HERE
		for (int i = 0; i < message.length(); i++)
		{
			fputc(message.charAt(i), stream);
		}
		return 0;
        }


	public static void main( String[] args ) {

		int element = 0;
		int count = 0;
		while (count < 3){
		
			element += ADD;
			count++;
		}
		writeline ("abc", System.err);
		newline(System.out);
		writeline ("def", System.out);
		newline(System.out);
		decout (123, System.out);
		newline(System.out);
		decout (0, System.out);
		newline(System.out);
		hexout (0xFEEDDAD, System.out);
		newline(System.out);
	}
}

File: hw1/hw1.c
/****************************************************************************

                                                        Darren Yau
                                                        CSE 12, W20
                                                        1/14/2020
                                                        Your cs12x login
                                Assignment One

File Name:      hw1.c
Description:    This program tests functions to display output strings and 
                numbers. 
****************************************************************************/
#include <stdio.h>

#define COUNT ((long) (sizeof (long) << 1 )) /* number of hex digits to 
                                                display */
#define DECIMAL 10	/* to indicate base 10 */
#define HEX 16		/* to indicate base 16 */
#define ADD 12		/* To be used in adding values in do while loop in main */

/* Function prototypes */
void baseout (long, long, FILE *);
void decout (unsigned long, FILE *);
void hexout (unsigned long, FILE *);
void newline (FILE *);
long writeline (const char *, FILE *);

/* array used for ASCII conversion */
const char digits[] = "0123456789ABCDEFGHIJKLMNOPQRSTUVWXYZ";

/*--------------------------------------------------------------------------
Function Name:         List the function name.
Purpose:               Say why the function exists in a sentence or two.
Description:           Describe the underlying algorithm of how the function
                       is implemented in a few sentences.
Input:                 A list of the parameters with brief explanation of each.
Result:                A list of possible outcomes when the function is called.
                       Describe return value.
--------------------------------------------------------------------------*/
void baseout (long number, long base, FILE * stream) {
	/* YOUR CODE GOES HERE */
	int index = 0;
	int remainder = 0;
	char array[] = {0,0,0,0,0,0,0,0};

	do
	{
		remainder = number % base;
		array[index] = digits[remainder];
		number /= base;
		index++;
	} while (number != 0);

	while (index >= 0)
	{
		fputc(array[index], stream);
		index--;
	}
}


/*--------------------------------------------------------------------------
Function Name:         List the function name.
Purpose:               Say why the function exists in a sentence or two.
Description:           Describe the underlying algorithm of how the function
                       is implemented in a few sentences.
Input:                 A list of the parameters with brief explanation of each.
Result:                A list of possible outcomes when the function is called.
                       Describe return value.
--------------------------------------------------------------------------*/
void decout (unsigned long number, FILE * stream) {
	/* YOUR CODE GOES HERE */
	baseout (number, DECIMAL, stream);
}


/*--------------------------------------------------------------------------
Function Name:         hexout
Purpose:               Prints a number in base 16 to the parameter FILE stream
Description:           Goal is achieved via delegating to the baseout function
Input:                 number:  the number to display
                       stream:  where to display, likely stdout or stderr
Result:                Number in base 16 is displayed.
                       No return value.
--------------------------------------------------------------------------*/
void hexout (unsigned long number, FILE * stream) {

	/* Output "0x" for hexidecimal. */
	writeline ("0x", stream);
	baseout (number, HEX, stream);
}


/*--------------------------------------------------------------------------
Function Name:         List the function name.
Purpose:               Say why the function exists in a sentence or two.
Description:           Describe the underlying algorithm of how the function
                       is implemented in a few sentences.
Input:                 A list of the parameters with brief explanation of each.
Result:                A list of possible outcomes when the function is called.
                       Describe return value.
--------------------------------------------------------------------------*/
void newline (FILE * stream) {
	/* YOUR CODE GOES HERE */
	fputc('\n', stream);
}


/*--------------------------------------------------------------------------
Function Name:         List the function name.
Purpose:               Say why the function exists in a sentence or two.
Description:           Describe the underlying algorithm of how the function
                       is implemented in a few sentences.
Input:                 A list of the parameters with brief explanation of each.
Result:                A list of possible outcomes when the function is called.
                       Describe return value.
--------------------------------------------------------------------------*/
long writeline (const char * message, FILE * stream) {
	/* YOUR CODE GOES HERE */
	int index = 0;
	while (message[index] != '\0')
	{
		fputc(message[index], stream);
		index++;
	}
	return 0;
}


int main (int argc, char *const* argv) {

	int element = 0;
	int count = 0;

	do {
		element += ADD;
		count++;

	}while(count < 3);

	writeline ("def", stdout);
	writeline ("abc", stderr);
	newline(stderr);
	newline(stdout);
	decout (123, stdout);
	newline(stdout);
	decout (0, stdout);
	newline(stdout);
	hexout (0xFEEDDAD, stdout);
	newline(stdout);
	return 0;
}
